modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryRequest.java

```diff
@@ -147,6 +152,10 @@ public class GridCacheQueryRequest extends GridCacheIdMessage implements GridCac
     /** */
     private AffinityTopologyVersion topVer;
 
+    /** Set of keys that must be skiped during iteration. */
+    @GridDirectCollection(KeyCacheObject.class)
+    private Collection<KeyCacheObject> skipKeys;
```

Comment:
There are several reason to filter keys on remote nodes:

1. Reduce network usage. We don't send values that will be filtered out.
2. In case transformer used, we invoke it on remote node and receive transformed value on local node. This reduce network usage, also.

Reply:
1. For example, you have 2 skipKeys and 10 nodes, in case of filtering on remote nodes you send additionaly 2 * 9 keys and receive 2 keys and 2 values less (but receive the same amount in case of insert only transactions). So, we can't be sure about reducing network usage. I think, in the typical usage, on the contrary, it will increase network usage.
2. Transformers can enlarge values as well.